pygfx/utils/load.py

```diff
@@ -5,72 +5,432 @@
+    apply_transforms : bool
+        Whether to apply the scene graph transforms directly to the meshes, by default True.
+
+    Returns
+    -------
+    meshes : list
+        A list of loaded meshes.
+
+    """
+    import trimesh  # noqa
+
+    # If this is a single object we can just convert and return it
     if isinstance(scene, trimesh.Trimesh):
         m = gfx.Mesh(
             gfx.geometry_from_trimesh(scene),
-            gfx.MeshPhongMaterial(),
+            gfx.material_from_trimesh(scene),
         )
         return [m]
-
+    # If this is a scene, we need to properly parse it
     elif isinstance(scene, trimesh.Scene):
+        # Scene consists of geometries (which in trimesh contain both
+        # the vertices/faces as well as the mateiral) and a scene graph.
+        # Each node in the graph consists of a geometry and a transform
+        # A geometry can be used by more than one node in the graph.
+        # Here we will convert each node in the graph to a pygfx.Mesh
+        # but in the future it might be better to use pygfx.InstancedMesh.
+
+        # Extract the geometries and materials
+        gfx_geometries, gfx_materials = objects_from_trimesh(scene, materials=materials)
+
+        # Generate a visual for each node
+        meshes = []
         for node_name in scene.graph.nodes_geometry:
             transform, geometry_name = scene.graph[node_name]
-            current = scene.geometry[geometry_name]
-            current.apply_transform(transform)
 
-        return [
-            gfx.Mesh(
-                gfx.geometry_from_trimesh(m),
-                gfx.material_from_trimesh(m.visual.material),
-            )
-            for m in scene.geometry.values()
-        ]
+            mesh = gfx.Mesh(gfx_geometries[geometry_name], gfx_materials[geometry_name])
+
+            if apply_transforms:
+                mesh.local.matrix = transform  # set the node's transform
+
+            meshes.append(mesh)
+        return meshes
     else:
         raise ValueError(f"Unexpected trimesh data: {scene.__class__.__name__}")
+
+
+def objects_from_trimesh(scene, materials=True):
+    """Extract geometries and materials from a trimesh scene.
+
+    Parameters
+    ----------
+    scene : trimesh.Scene
+        The scene to convert.
+    materials : bool
+        Whether to import materials. If False, a default material will be created.
+
+    Returns
+    -------
+    gfx_geometries : dict
+        A dictionary of geometry objects. Keys are the names of the geometries.
+    gfx_materials : dict
+        A dictionary of material objects. Keys are the names of the geometries.
+
+    """
+    gfx_geometries = {}
+    gfx_materials = {}
+    for name, mesh in scene.geometry.items():
+        gfx_geometries[name] = gfx.geometry_from_trimesh(mesh)
+        # If mesh has a material, convert it
+        if hasattr(mesh.visual, "material") and materials:
+            gfx_materials[name] = gfx.material_from_trimesh(mesh.visual.material)
```

Comment:
Is it possible to tell if two trimesh materials are identical/equivalent? Same question for geometry.

If possible, you can share the geometry and material objects between WorldObjects and reduce VRAM usage.

Reply:
Yes, you are right: at least materials can be identical. I assume that can also be the case for geometries but I haven't seen that in the wild.

I'm not sure about about equivalent though - it doesn't look like trimesh materials/objects implement any special `__eq__` methods.

With 2e4b649 this is now de-duplicating materials and geometries based on identity but not equivalence.